domain/storage/service/storage.go

```diff
@@ -0,0 +1,145 @@
+	poolDetails, err := s.st.GetStoragePoolByName(ctx, arg.Pool)
+	if errors.Is(err, storageerrors.PoolNotFoundError) {
+		poolDetails = storage.StoragePoolDetails{
+			Name:     arg.Pool,
+			Provider: arg.Pool,
+			Attrs:    map[string]string{},
+		}
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+	registry, err := s.registryGetter.GetStorageRegistry(ctx)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	provider, err := registry.StorageProvider(internalstorage.ProviderType(poolDetails.Provider))
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	details, err := s.st.GetModelDetails()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	resourceTags := map[string]string{
+		tags.JujuModel:      details.ModelUUID,
+		tags.JujuController: details.ControllerUUID,
+	}
+	filesystemInfo := storage.FilesystemInfo{Pool: arg.Pool}
+
+	// If the storage provider supports filesystems, import the filesystem,
+	// otherwise import a volume which will back a filesystem.
+	invalidatorFunc, err := credentialInvalidatorGetter()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	callCtx := envcontext.WithCredentialInvalidator(ctx, invalidatorFunc)
+
+	var attr map[string]any
+	if len(poolDetails.Attrs) > 0 {
+		attr = transform.Map(poolDetails.Attrs, func(k, v string) (string, any) { return k, v })
+	}
+	cfg, err := internalstorage.NewConfig(poolDetails.Name, internalstorage.ProviderType(poolDetails.Provider), attr)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	if provider.Supports(internalstorage.StorageKindFilesystem) {
+		filesystemSource, err := provider.FilesystemSource(cfg)
+		if err != nil {
+			return "", errors.Capture(err)
+		}
+		filesystemImporter, ok := filesystemSource.(internalstorage.FilesystemImporter)
+		if !ok {
+			return "", errors.Errorf(
+				"importing filesystem with storage provider %q not supported",
+				cfg.Provider(),
+			).Add(coreerrors.NotSupported)
+		}
+		info, err := filesystemImporter.ImportFilesystem(callCtx, arg.ProviderId, resourceTags)
+		if err != nil {
+			return "", errors.Errorf("importing filesystem: %w", err)
+		}
+		filesystemInfo.FilesystemId = arg.ProviderId
+		filesystemInfo.Size = info.Size
+	} else {
+		volumeSource, err := provider.VolumeSource(cfg)
+		if err != nil {
+			return "", errors.Capture(err)
+		}
+		volumeImporter, ok := volumeSource.(internalstorage.VolumeImporter)
+		if !ok {
+			return "", errors.Errorf(
+				"importing volume with storage provider %q not supported",
+				cfg.Provider(),
+			).Add(coreerrors.NotSupported)
+		}
+		info, err := volumeImporter.ImportVolume(callCtx, arg.ProviderId, resourceTags)
+		if err != nil {
+			return "", errors.Errorf("importing volume: %w", err)
+		}
+		filesystemInfo.BackingVolume = &internalstorage.VolumeInfo{
+			HardwareId: info.HardwareId,
+			WWN:        info.WWN,
+			Size:       info.Size,
+			VolumeId:   info.VolumeId,
+			Persistent: info.Persistent,
+		}
+		filesystemInfo.Size = info.Size
+	}
```

Comment:
This should be a couple of methods. It's doing way too much.

Reply:
The whole block of code was copied directly from the facade. I can clean up a bit.